### ISSUE TITLE
FlashAttention3 producing NaN output when NaN exist in parts of input data that it should not be reading

### ISSUE THREAD

**Issue by tdoublep** [Dao-AILab/flash-attention#1974]
We've hit an issue in vLLM when supporting hybrid models (e.g., attention + mamba2) when the mamba state can have different dtype to attention state (KV cache). In vLLM V1, the mamba state and KV cache is managed in a unified way and they both share the same underlying memory but using different views. This can lead to a situation when the bits in the KV tensor look like NaN in bf16 (although they were valid numbers in fp32). 

What is strange is that these NaN's seem to affect FA3 in a way that we do not understand. It looks like the presence of the NaN in the KV tensor can lead to getting NaNs in the output even though theoretically the kernel should not be reading these values. We have created the following reproducer:
```python
import torch
from flash_attn_interface import _flash_attn_forward

N_TOKENS = 107
BLOCK_SIZE = 256
N_BLOCKS = 1
N_HEADS = 32
N_KV_HEADS = 8
HEAD_DIM = 128

torch.set_default_dtype(torch.bfloat16)
torch.set_default_device('cuda:0')

query = torch.rand(N_TOKENS, N_HEADS, HEAD_DIM)
key_cache = torch.rand(N_BLOCKS, BLOCK_SIZE, N_KV_HEADS, HEAD_DIM)
value_cache = torch.rand(N_BLOCKS, BLOCK_SIZE, N_KV_HEADS, HEAD_DIM)
cu_seqlens_q = torch.tensor([0, N_TOKENS], dtype=torch.int32)
seqused_k = torch.tensor([N_TOKENS], dtype=torch.int32)
block_table = torch.tensor([[0]], dtype=torch.int32)

# fill key_cache and value_cache with nans *outside* of
# the range where they should be read
key_cache[:, N_TOKENS:].fill_(float('nan'))
value_cache[:, N_TOKENS:].fill_(float('nan'))

output = torch.empty_like(query)

_flash_attn_forward(
    query,
    key_cache,
    value_cache,
    None, # k_new,
    None, # v_new
    None, # q_v
    output,
    cu_seqlens_q,
    None, # cu_seqlens_k,
    None, # cu_seqlens_k_new
    None, # seqused_q,
    seqused_k,
    N_TOKENS, # max_seqlen_q
    N_TOKENS, # max_seqlen_k
    block_table,
    None,             # kv_batch_idx
    None,             # leftpad_k
    None, None, None, # rotary_cos, rotary_sin, seqlens_rotary
    None, None, None, # q_descale, k_descale, v_descale,
    1.0, # softmax_scale
    True, # causal
    window_size=(-1,-1),
    attention_chunk=0,
    softcap=0,
    rotary_interleaved=True,
    scheduler_metadata=None,
    num_splits=0,
    pack_gqa=None,
    sm_margin=0,
)

assert not torch.isnan(output).any()
```
The kernel shouldn't actually be reading anything from `key_cache[:, N_TOKENS:]` or `value_cache[:, N_TOKENS:]` but the presence of NaNs there seems to cause the above test to fail. 

I also observed that the issue goes away if I set `NUM_HEADS=NUM_KV_HEADS` (e.g., disable GQA). My hypothesis is reading beyond `N_TOKENS` due to some tiling/cache line alignment, and the NaN's are somehow not being masked out correctly.

**Comment by tridao** [comment]
Yes, for varlen the kernel will read the V beyond the NTOKENS (it reads blocks of e.g. 128 tokens at a time). Typically this is ok because we then multiply by the attn probability, which will be zero beyond the NTOKENS. But when there are inf or NaN values beyond NTOKENS, multiplying by zero would produce NaN.
The "right" approach is for the kernel to update the TMA tensormap to avoid reading beyond NTOKENS, but that's a bit complicated so we haven't implemented that. At some point we will.

**Comment by vadiklyutiy** [comment]
Faced to similar problem with Qwen3.5 

Also if hybrid cache just outstand the issue I think the issue is possible in some common case. 

FA3 use mul by 0 to mask not used values. 
To be sure that its correct we must guarantee that no NaN. In general case masking by mul by 0 is incorrect.

Seems even common full attn can produce NaN for some corner cases. It is likely ok to produce NaN(and garbage tokens) for one specific request, but with kernels that not tolerant to NaN, we propagate this problem to another good requests with subsequent garbage for all requests.

**Comment by vadiklyutiy** [comment]
@tridao 
Could you also clarify does FA2 has the same behavior and not tolerate to NaN?
